### PR TITLE
[3320] support membership property

### PR DESCRIPTION
**Pull request by kanat**
### 🎯 Goal

Support `Channel.membership` property

### 🛠 Implementation details

[Channel membership property lifecycle](https://www.notion.so/stream-wiki/Android-Channel-membership-property-lifecycle-fc01903e4c7a4c7cae2498a061aa1375)

### 🧪 Testing

Check databases, table `stream_chat_channel_state`, column `membership` 

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] ~~New code is covered by unit tests~~
- [ ] ~~Comparison screenshots added for visual changes~~
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

**Review by bychkovdmitry — COMMENTED** [review]
Some tests don't compile anymore.

**Review by bychkovdmitry — COMMENTED** [review]
I've added logging to `ChannelListView.setChannels(channels: List<ChannelListItem>)`. When I enter a channel and then return back to the channel list, the channel shows null `membership`. Could you have a look?

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,8 @@
 ### ⬆️ Improved
 
 ### ✅ Added
+- Added `Channel.membership` property. [#3320](https://github.com/GetStream/stream-chat-android/pull/3320)
+- Added `NotificationAddedToChannelEvent.member` property. [#3320](https://github.com/GetStream/stream-chat-android/pull/3320)
```
[the remark this answers is not in the document]
Comment (kanat) [reply CHANGELOG.md:20]: Ah, it seems to be lost after rebasing 🙈

**Comment by bychkovdmitry** [comment]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Review by bychkovdmitry — APPROVED** [review]
LGTM!

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt [hunk 2]
```diff
@@ -52,24 +52,114 @@ internal fun Channel.updateLastMessage(message: Message) {
...
+    return this
+}
+
+/**
+ * Updates member in the [Channel.members].
+ *
+ * @param memberUserId Updated member user id.
+ * @param banned Shows whether a user is banned or not in this channel.
+ */
+internal fun Channel.updateMember(memberUserId: String?, banned: Boolean): Channel {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt:104]: I think we could change its name + docs as it only updates `member.banned`

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt [hunk 3]
```diff
@@ -52,24 +52,114 @@ internal fun Channel.updateLastMessage(message: Message) {
...
+    }
+    return this
+}
+
+/**
+ * Sets [Channel.membership] to [member] if [member.user.id] equals to [Channel.membership.user.id].
+ *
+ * @param member Updated member.
+ */
+internal fun Channel.updateMembership(member: Member?): Channel {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt:133]: What's the difference between this and `addMembership`? 🤔 
`Channel.membership` is only applicable for the current user, right? Maybe we can merge those two methods

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt [hunk 4]
```diff
@@ -52,24 +52,114 @@ internal fun Channel.updateLastMessage(message: Message) {
...
+    return this
+}
+
+/**
+ * Sets [Channel.membership.banned] to [banned] if [memberUserId] equals to [membership.user.id].
+ *
+ * @param memberUserId Member user id.
+ * @param banned Shows whether a user is banned or not in this channel.
+ */
+internal fun Channel.updateMembership(memberUserId: String?, banned: Boolean): Channel {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt:146]: Should we change its name as it only updates `membership.banned`? 🤔

File: stream-chat-android-offline/detekt-baseline.xml [hunk 5]
```diff
@@ -132,6 +132,8 @@
     <ID>NestedBlockDepth:FilterObjectConverter.kt$private fun Map.Entry&lt;String, Any>.toFilterObject(): FilterObject</ID>
     <ID>NestedBlockDepth:UploadAttachmentsWorker.kt$UploadAttachmentsWorker$suspend fun uploadAttachmentsForMessage( channelType: String, channelId: String, messageId: String, ): Result&lt;Unit></ID>
     <ID>PrintStackTrace:UploadAttachmentsWorker.kt$UploadAttachmentsWorker$e</ID>
+    <ID>ReturnCount:Channel.kt$ internal fun Channel.addMember(member: Member?): Channel</ID>
```
Comment (adasiewiczr) [inline stream-chat-android-offline/detekt-baseline.xml:135]: Why do we need to add those two methods here/

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt (hunk 3 above)
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt:132]: Inside the `Channel.updateMembership` we rely ing on `membership` existence , while in `addMembership` we assume that `membership` is most likely `null`.

File: stream-chat-android-offline/detekt-baseline.xml (hunk 5 above)
Comment (kanat) [reply stream-chat-android-offline/detekt-baseline.xml:132]: Because of having several `return` inside the method.
Dunno why it's restricted.

**Review by adasiewiczr — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-offline/detekt-baseline.xml (hunk 5 above)
Comment (adasiewiczr) [reply stream-chat-android-offline/detekt-baseline.xml:132]: @JcMinarro what do you think?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt (hunk 3 above)
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/internal/Channel.kt:135]: makes sence

**Pull request merged**